open_lm/file_utils.py

```diff
@@ -93,33 +94,88 @@ def get_metadata_file(path):
+
+    Chunks are groups of shards with samples roughly equal to the number of samples
+    that will be seen during training. This function uses the dataset manifest
+    to split the shards into chunks, and assign shards to each chunk.
+    """
     metadata = get_metadata_file(path)
     shard_list = []
     curr_shard_list = []
     chunk_count_list = []
     curr_chunk_count = 0
-    real_chunk_count = 0
     for m in metadata:
         curr_chunk_count += m['num_chunks']
-        real_chunk_count += m['num_chunks']
         curr_shard_list.append(m['shard'])
         if curr_chunk_count >= num_samples:
             shard_list.append(curr_shard_list)
-            curr_chunk_count = curr_chunk_count - num_samples
+            chunk_count_list.append(curr_chunk_count)
             curr_shard_list = []
-            chunk_count_list.append(real_chunk_count)
-            real_chunk_count = 0
+            curr_chunk_count = 0
     
-    return shard_list[epoch % len(shard_list)], chunk_count_list[epoch % len(chunk_count_list)]
-
-def get_string_for_epoch(num_samples, epoch, path):
-    shard_list, num_samples = get_shards_for_epoch(num_samples, epoch, path)
-    shard_root = '/'.join(path.split('/')[:-1]) + '/shard_'
-    shard_string = shard_root + '{' + shard_list[0] + '..' + shard_list[-1] + '}.tar'
-    if path.startswith('s3'):
-        shard_string = f'pipe:aws s3 cp {shard_string} -'
-    return shard_string, num_samples
+    # Append remaining shards
+    if len(curr_shard_list) > 0:
+        shard_list.append(curr_shard_list)
+        chunk_count_list.append(curr_chunk_count)
+
+    return shard_list[chunk % len(shard_list)], chunk_count_list[chunk % len(chunk_count_list)]
+
+
+def enough_shards(shard_lists, min_shards_needed):
+    for sl in shard_lists:
+        if len(sl) < min_shards_needed:
+            return False
+    return True
+
+
+def enough_samples(num_samples_per_source, needed_samples_per_source):
+    for i, number in enumerate(num_samples_per_source):
+        if number < needed_samples_per_source[i]:
+            return False
+    return True
+
+
+def source_exhausted(paths, shard_list_per_source):
+    for i, source in enumerate(paths):
+        data = get_metadata_file(source)
+        if len(data) < len(shard_list_per_source[i]):
+            return True
+    return False
+
+
+def get_string_for_epoch(num_samples, starting_chunk, paths, weights, min_shards_needed):
+    if weights is None:
+        weights = [1.0 / len(paths) for _ in range(len(paths))]
+    needed_samples_per_source = [int(np.ceil(weights[i] * num_samples / sum(weights))) for i in range(len(weights))]
+    shard_strings_per_source = []
+    next_chunk = starting_chunk
+    shard_list_per_source = [[] for _ in range(len(paths))]
+    num_samples_per_source = [0 for _ in range(len(paths))]
+    while not enough_shards(shard_list_per_source, min_shards_needed) or not enough_samples(num_samples_per_source, needed_samples_per_source):
```

Comment:
I'm a little confused here; why do we need a loop here if get_shards_for_chunk also has a loop that packs shards until we have >= needed_samples_per_epoch? Is it only for the the last chunk, where we may have fewer samples than needed?

If so, I would have thought we should handle this in get_shards_for_chunk, but I suppose it's possible that we have one data source with not enough samples, and another with more than enough samples, and things work out for us serendipitously. If my understanding is correct, I think no changes needed, but might be good to drop a comment to this effect.